executor/src/witgen/mod.rs

```diff
@@ -549,6 +567,35 @@ impl<'a, T: FieldElement> FixedData<'a, T> {
                 v.get(row as usize).cloned()
             })
     }
+
+    fn current_stage_witnesses(&self) -> impl Iterator<Item = PolyID> + '_ {
+        self.witness_cols
+            .iter()
+            .filter(|(_, col)| col.stage <= self.stage as u32)
```

Comment:
Doesn't this filtering already happen in the constructor?

Reply:
And shouldn't it be `== self.stage`?

Reply:
> Doesn't this filtering already happen in the constructor?

No

> And shouldn't it be `== self.stage`?

I mean it's only currently used by the `MachineExtractor`, which is only run in stage 0, so it doesn't matter in practice. But I think it makes more sense to have a function that returns all witnesses relevant in the current stage, which would include previous-stage witnesses. Maybe I should rename it to `witnesses_for_current_stage()`?

Reply:
`witnesses_until_current_stage`

Reply:
Done!